fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingByIndex.java

```diff
@@ -182,20 +190,98 @@ private void maybeLogBuildProgress(SubspaceProvider subspaceProvider, byte[] ret
 
         final IndexMaintainer maintainer = store.getIndexMaintainer(index);
 
+        // this should never happen. But it makes the compiler happy
+        validateOrThrowEx(srcIndex != null, "source index is null");
         // idempotence - We could have verified it at the first iteration only, but the repeating checks seem harmless
         validateOrThrowEx(maintainer.isIdempotent(), "target index is not idempotent");
+        // readability - This method shouldn't block if one has already opened the record store (as we did)
+        validateOrThrowEx(store.isIndexReadable(srcIndex), "source index is not readable");
 
-        // this should never happen. But it makes the compiler happy
-        validateOrThrowEx(srcIndex != null, "source index is null");
+        RangeSet rangeSet = new RangeSet(store.indexRangeSubspace(index));
+        AsyncIterator<Range> ranges = rangeSet.missingRanges(store.ensureContextActive(), startBytes, endBytes).iterator();
+
+        final ExecuteProperties.Builder executeProperties = ExecuteProperties.newBuilder()
+                .setIsolationLevel(IsolationLevel.SNAPSHOT)
+                .setReturnedRowLimit(getLimit()); // always respectLimit in this path
+        final ScanProperties scanProperties = new ScanProperties(executeProperties.build());
+
+        return ranges.onHasNext().thenCompose(hasNext -> {
+            if (Boolean.FALSE.equals(hasNext)) {
+                return AsyncUtil.READY_FALSE; // no more missing ranges - all done
+            }
+            Range range = ranges.next();
+            Tuple rangeStart = Arrays.equals(range.begin, START_BYTES) ? null : Tuple.fromBytes(range.begin);
+            Tuple rangeEnd = Arrays.equals(range.end, END_BYTES) ? null : Tuple.fromBytes(range.end);
+
+            RecordCursor<FDBIndexedRecord<Message>> cursor =
+                    store.scanIndexRecords(srcIndex, IndexScanType.BY_VALUE, TupleRange.between(rangeStart, rangeEnd), range.begin, scanProperties);
```

Comment:
It feels wrong to me that this is using `range.begin` as the continuation into `scanIndexRecords` here. It's typically not good practice to use a continuation from anything other than `getContinuation` or to treat it as anything other than an opaque byte array. For example, we reserve the right to, say, wrap the continuation in a protobuf at some point in the future, at which point this will break mysteriously. So, I think this should be `null` here, and then I have a suggestion below about how to not grab the continuation any more.

Reply:
When I used `null`, the cursor included the `range.begin` - which caused duplication. Is there any other way to exclude it?

Reply:
digging a little more into the code - 
`range.begin`, if not null, is produced by `lastResult.getContinuation()` and should be used as continuation arg.
`range.end`, if not null, is either defined by the caller or as a missingRange - making a top boundary, therefore used as range arg. 
I've set the range to '(null, range.end)` and added a clarifying comment. 
Now the question is whether or not a continuation should be a legal range end at ` rangeSet.insertRange`. Should it?

Reply:
(adding a thought) 
I believe it is safe to assume that `getContinuation` return ordinal keys, and therefore can be used in `rangeSet`. 
When the time comes, though, we should remember to use continuation keys as range for parallel indexing.